players/g3_player.py

```diff
@@ -207,51 +208,107 @@ def play(self, unit_id: List[List[str]], unit_pos: List[List[Point]], map_states
+                if neighbor != curr_pt and self._on_border((neighbor)):
+                    self._trace_border(neighbor, border_pts)
+
+    def _on_border(self, pt):
+        """Check if given point is on the border"""
+        # currently ignoring disputed cells
+        if self.map_states[pt] != self.us: 
+            return False
+
+        xmax, ymax = self.map_states.shape
+        neighbors = np.array([[min(pt[0]+1, xmax-1), pt[1]], 
+            [max(pt[0]-1, 0), pt[1]],
+            [pt[0], min(pt[1]+1, ymax-1)],
+            [pt[0], max(pt[1]-1, 0)]])
+        return any(self.map_states[neighbors[:, 0], neighbors[:, 1]] != self.us)
+
+    def _explore(self, scout_unit, enemy_clusters, ally_clusters):
         homebase_force = inverse_force((scout_unit - self.homebase).reshape(1, 2))
         force = exploration_force(scout_unit, enemy_clusters, ally_pts=ally_clusters) \
-            + border_repulsion(scout_unit, xmax=len(self.map_states), ymax=len(self.map_states[0])) \
-            + SCOUT_ALLY_SCALE * homebase_force
+            + SCOUT_BORDER_SCALE * border_repulsion(scout_unit, xmax=self.map_states.shape[0], ymax=self.map_states.shape[1]) \
+            + SCOUT_HOMEBASE_SCALE * homebase_force
         return np.array([1, np.arctan2(force[1], force[0])])
 
+    def _get_clusters(self, ally_units):
+        # keep this incase we need it later
+        # enemy_k = min(50, math.ceil(self.enemy_units.shape[0]/2))
+        # enemy_clusters = KMeans(n_clusters=enemy_k).fit(self.enemy_units).cluster_centers_
+        # ally_k = min(15, math.ceil(ally_units.shape[0]/2))
+        # ally_clusters = KMeans(n_clusters=ally_k).fit(ally_units).cluster_centers_
+
+        # change to random selection to speed up
+        enemy_clusters = self.enemy_units[np.random.choice(np.arange(self.enemy_units.shape[0]), min(self.enemy_units.shape[0], 50), replace=False)]
+        ally_clusters = ally_units[np.random.choice(np.arange(ally_units.shape[0]), min(ally_units.shape[0], 15), replace=False)]
+        return enemy_clusters, ally_clusters
+
     def move_scouts(self, scout_ids):
         scout_units = self.our_units[scout_ids]
         scout_moves = np.zeros_like(scout_units, dtype=float)
         # safety check
         ally_units = np.delete(self.our_units, scout_ids, axis=0)
         ally_dist = ((scout_units.reshape(-1, 1, 2) - ally_units.reshape(1, -1, 2)) ** 2).sum(axis=2)
         min_ally_id = ally_dist.argmin(axis=1)
-        min_enemy_dist = ((scout_units.reshape(-1, 1, 2) - self.enemy_units.reshape(1, -1, 2)) ** 2).sum(axis=2).argmin(axis=1)
+        min_enemy_dist = ((scout_units.reshape(-1, 1, 2) - self.enemy_units.reshape(1, -1, 2)) ** 2).sum(axis=2).min(axis=1)
+
+        enemy_clusters, ally_clusters = self._get_clusters(ally_units)
         for i in range(scout_units.shape[0]):
-            if ally_dist[i, min_ally_id[i]] >= min_enemy_dist[i]:
+            if ally_dist[i, min_ally_id[i]] >= min_enemy_dist[i] * 2:
```

Comment:
curious why a value of 2 is used.

Reply:
No specific reason actually. 1 is guaranteed to be safe (all directions considered; e.g., enemy and ally comes form exact same direction) but to be more aggressive 2 seems to work fine.